stdlib/lexing.mli

```diff
@@ -108,6 +108,15 @@ val from_function : ?with_positions:bool -> (bytes -> int -> int) -> lexbuf
    starting at index 0, and return the number of bytes
    provided. A return value of 0 means end of input. *)
 
+val set_position : lexbuf -> position -> unit
+(** Set the initial tracked input position for [lexbuf] to custom value.
+   Does not set file (use [set_file] for it). Can be used to adjust location
+   information without need to traverse AST. *)
+
+val set_filename: lexbuf -> string -> unit
+(** [set_filename lexbuf file] sets the filename to [file] in
+    [lexbuf]. *)
```

Comment:
```
@since 4.10
```

Reply:
The comment is misleading if the function is (erroneously) called after lexing has started, since it only updates `lex_curr_p`. Given that this function is supposed to be called *before* lexing begins, wouldn't it be better embedded as an optional argument in the `from_` functions?

Reply:
> The comment is misleading if the function is (erroneously) called after lexing has started, since it only updates lex_curr_p.

Thanks! We can update `lex_start_p` as well. It is very unlikely that this overhead creates performance problems.

> wouldn't it be better embedded as an optional argument in the from_ functions?

1. I wanted to have `Lexing` and `Sedlexing` be as compatible as possible.
2. If I understand it right, both functions can be called during lexing - this will shift the positions of next tokens. To make this more reliable, we can alter both `lex_start_p` and `lex_curr_p`.

Does this plan (implement two proposed functions, shift both `lex_start_p` and `lex_curr_p`) look good for you?

Reply:
For the record, I'm not fond of adding optional arguments to `from_*` functions. I think that APIs resulting from this process and heavyweight and not-so-pleasant to use.

I would prefer doing `with_pos p (Lexing.from_string s)` over `Lexing.from_string ~pos:p s`. Given that lexing buffers are mutable, a `set_position`  mutator feels reasonable and consistent to me.